File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/ListHabitsController.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list;

import android.support.annotation.*;

import org.isoron.androidbase.activities.*;
import org.isoron.uhabits.R;
import org.isoron.uhabits.activities.habits.list.views.*;
import org.isoron.uhabits.core.models.*;
import org.isoron.uhabits.core.tasks.*;
import org.isoron.uhabits.core.ui.screens.habits.list.*;
import org.isoron.uhabits.tasks.*;

import java.io.*;

import javax.inject.*;

@ActivityScope
public class ListHabitsController
    implements HabitCardListController.HabitListener
{
    @NonNull
    private final ListHabitsBehavior behavior;

    @NonNull
    private final ListHabitsScreen screen;

    @NonNull
    private final HabitCardListAdapter adapter;

    @NonNull
    private final TaskRunner taskRunner;

    private ImportDataTaskFactory importTaskFactory;

    private ExportDBTaskFactory exportDBFactory;

    @Inject
    public ListHabitsController(@NonNull ListHabitsBehavior behavior,
                                @NonNull HabitCardListAdapter adapter,
                                @NonNull ListHabitsScreen screen,
                                @NonNull TaskRunner taskRunner,
                                @NonNull ImportDataTaskFactory importTaskFactory,
                                @NonNull ExportDBTaskFactory exportDBFactory)
    {
        this.behavior = behavior;
        this.adapter = adapter;
        this.screen = screen;
        this.taskRunner = taskRunner;
        this.importTaskFactory = importTaskFactory;
        this.exportDBFactory = exportDBFactory;
    }

    public void onEdit(@NonNull Habit habit, long timestamp)
    {
        behavior.onEdit(habit, timestamp);
    }

    public void onExportCSV()
    {
        behavior.onExportCSV();
    }

    public void onExportDB()
    {
        taskRunner.execute(exportDBFactory.create(filename ->
        {
            if (filename != null) screen.showSendFileScreen(filename);
            else screen.showMessage(R.string.could_not_export);
        }));
    }

    @Override
    public void onHabitClick(@NonNull Habit h)
    {
        behavior.onClickHabit(h);
    }

    @Override
    public void onHabitReorder(@NonNull Habit from, @NonNull Habit to)
    {
        behavior.onReorderHabit(from, to);
    }

    public void onImportData(@NonNull File file,
                             @NonNull OnFinishedListener finishedListener)
    {
        taskRunner.execute(importTaskFactory.create(file, result ->
        {
            switch (result)
            {
                case ImportDataTask.SUCCESS:
                    adapter.refresh();
                    screen.showMessage(R.string.habits_imported);
                    break;

                case ImportDataTask.NOT_RECOGNIZED:
                    screen.showMessage(R.string.file_not_recognized);
                    break;

                default:
                    screen.showMessage(R.string.could_not_import);
                    break;
            }

            finishedListener.onFinish();
        }));
    }

    public void onInvalidEdit()
    {
        screen.showMessage(R.string.long_press_to_edit);
    }

    @Override
    public void onInvalidToggle()
    {
        screen.showMessage(R.string.long_press_to_toggle);
    }

    public void onRepairDB()
    {
        behavior.onRepairDB();
    }

    public void onSendBugReport()
    {
        behavior.onSendBugReport();
    }

    public void onStartup()
    {
        behavior.onStartup();
    }

    @Override
    public void onToggle(@NonNull Habit habit, long timestamp)
    {
        behavior.onToggle(habit, timestamp);
    }

    public interface OnFinishedListener
    {
        void onFinish();
    }
}

File: ios/Application/ListHabitsController.swift
/*
 * Copyright (C) 2016-2019 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import UIKit

class ListHabitsCell : UITableViewCell {
    var ring: ComponentView
    var label = UILabel()
    var buttons: [ComponentView] = []
    var theme = LightTheme()
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        ring = ComponentView(frame: CGRect(), component: nil)
        
        super.init(style: .default, reuseIdentifier: reuseIdentifier)
        let size = CGFloat(theme.checkmarkButtonSize)
        
        let stack = UIStackView(frame: contentView.frame)
        stack.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        stack.backgroundColor = .red
        stack.axis = .horizontal
        stack.distribution = .fill
        stack.alignment = .center
        contentView.addSubview(stack)

        ring.backgroundColor = .white
        ring.widthAnchor.constraint(equalToConstant: size * 0.75).isActive = true
        ring.heightAnchor.constraint(equalToConstant: size).isActive = true
        stack.addArrangedSubview(ring)
        
        label.backgroundColor = .white
        label.heightAnchor.constraint(equalToConstant: size).isActive = true
        stack.addArrangedSubview(label)
        
        for _ in 1...4 {
            let btn = ComponentView(frame: frame, component: nil)
            btn.backgroundColor = .white
            btn.widthAnchor.constraint(equalToConstant: size).isActive = true
            btn.heightAnchor.constraint(equalToConstant: size).isActive = true
            buttons.append(btn)
            stack.addArrangedSubview(btn)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError()
    }
    
    func setColor(_ color: Color) {
        label.textColor = color.uicolor
        ring.component = Ring(color: color,
                              percentage: Double.random(in: 0...1),
                              thickness: 2.5,
                              radius: 7,
                              theme: theme,
                              label: false)
        ring.setNeedsDisplay()
        let isNumerical = Int.random(in: 1...4) == 1
        for btn in buttons {
            if isNumerical {
                btn.component = NumberButton(color: color,
                                             value: Double.random(in: 0...5000),
                                             threshold: 2000,
                                             units: "steps",
                                             theme: theme)
            } else {
                btn.component = CheckmarkButton(value: Int32.random(in: 0...2),
                                                color: color,
                                                theme: theme)
            }
            btn.setNeedsDisplay()
        }
    }
}

class ListHabitsController: UITableViewController {
    var backend: Backend
    var habits: [[String: Any]]
    var theme: Theme
    
    required init?(coder aDecoder: NSCoder) {
        fatalError()
    }
    
    init(withBackend backend:Backend) {
        self.backend = backend
        self.habits = backend.getHabitList()
        self.theme = backend.theme
        super.init(nibName: nil, bundle: nil)
    }

    override func viewDidLoad() {
        self.title = "Habits"
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add,
                                                                 target: self,
                                                                 action: #selector(self.onCreateHabitClicked))
        tableView.register(ListHabitsCell.self, forCellReuseIdentifier: "cell")
        tableView.backgroundColor = theme.headerBackgroundColor.uicolor
    }

    @objc func onCreateHabitClicked() {
        self.navigationController?.pushViewController(EditHabitController(), animated: true)
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return habits.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let row = indexPath.row
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! ListHabitsCell
        let color = theme.color(paletteIndex: habits[row]["color"] as! Int32)
        cell.label.text = habits[row]["name"] as? String
        cell.setColor(color)
        return cell
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let component = HabitListHeader(today: LocalDate(year: 2019, month: 3, day: 24),
                                        nButtons: 4,
                                        theme: theme,
                                        fmt: IosLocalDateFormatter(),
                                        calc: IosLocalDateCalculator())
        return ComponentView(frame: CGRect(x: 0, y: 0, width: 100, height: CGFloat(theme.checkmarkButtonSize)),
                             component: component)
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return CGFloat(theme.checkmarkButtonSize)
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return CGFloat(theme.checkmarkButtonSize) + 1
    }
    
}
